Enable ability to read solvables from file by prefixing name with @

mmathesius opened the pull request:
Fixes: https://github.com/fedora-modularity/fus/issues/24

Signed-off-by: Merlin Mathesius <mmathesi@redhat.com>

igor-raits commented on main.c:
```diff
@@ -612,6 +612,23 @@ apply_excludes (Pool       *pool,
+static void
+add_solvable_to_queue(Pool *pool, Queue *selection, const char *solvable, int flags, Queue *solvq)
+{
+ selection_make (pool, selection, solvable, flags);
```
indent should be 2

igor-raits commented on main.c:
```diff
@@ -612,6 +612,23 @@ apply_excludes (Pool       *pool,
 }
 
+static void
+add_solvable_to_queue(Pool *pool, Queue *selection, const char *solvable, int flags, Queue *solvq)
```
arguments should be aligned vertically.

igor-raits commented on main.c:
```diff
@@ -789,18 +806,27 @@ main (int   argc,
+          gchar *content;
+          if (!g_file_get_contents (filename, &content, &length, NULL))
+            {
+              g_warning ("Unable to read file %s", filename);
```
this is error, really

igor-raits commented on main.c:
```diff
@@ -789,18 +806,27 @@ main (int   argc,
+              g_warning ("Unable to read file %s", filename);
+              continue;
+            }
+          g_auto(GStrv) sv = g_strsplit_set(content, " \t\n", 0);
```
no, make it line by line

igor-raits commented on main.c:
```diff
@@ -789,18 +806,27 @@ main (int   argc,
+          gchar *filename = *solvable+1;
+          gsize length;
+          gchar *content;
+          if (!g_file_get_contents (filename, &content, &length, NULL))
```
do not read full file into memory

igor-raits replied:
no error propagation

igor-raits commented on main.c:
```diff
@@ -789,18 +806,27 @@ main (int   argc,
         {
-          g_warning ("Nothing matches '%s'", *solvable);
-          continue;
+          gchar *filename = *solvable+1;
```
you can use *solvable++ instead

igor-raits commented on main.c:
```diff
@@ -789,18 +806,27 @@ main (int   argc,
-          continue;
+          gchar *filename = *solvable+1;
+          gsize length;
+          gchar *content;
```
memory leak

mmathesius commented:
@ignatenkobrain Thank you for reviewing my PR. I have revised my PR based on your feedback, as well as rebasing onto the current master which includes bug fixes.

igor-raits commented on main.c:
```diff
@@ -612,6 +613,70 @@ apply_excludes (Pool       *pool,
 }
 
+static void
+add_solvable_to_queue(const char *solvable,
```
Space before parentheses

igor-raits commented on main.c:
```diff
@@ -612,6 +613,70 @@ apply_excludes (Pool       *pool,
+}
+
+static void
+add_solvables_from_file_to_queue(const char *filename,
```
Same here

igor-raits commented on main.c:
```diff
@@ -612,6 +613,70 @@ apply_excludes (Pool       *pool,
+    g_set_error (&err,
+                 G_FILE_ERROR, g_file_error_from_errno (errno),
+                 "Unable to open file '%s': %s", filename, g_strerror (errno));
+    exiterr (err);
```
Never to that.

mmathesius replied:
Never do what?

mmathesius replied:
Were you looking for something more like?:
```
  GIOChannel *ch = g_io_channel_new_file (filename, "r", &err);
  if (ch == NULL)
  {
    g_printerr ("Unable to open file '%s': %s\n", filename, err->message);
    exit (EXIT_FAILURE);
  }
```

igor-raits commented on main.c:
```diff
@@ -612,6 +613,70 @@ apply_excludes (Pool       *pool,
+
+    if (tpos)
+      content[tpos] = '\0';
+    g_auto(GStrv) sv = g_strsplit_set (content, " \t", 0);
```
One line - one solvable.

mmathesius commented:
@ignatenkobrain Please review my latest updated commit. Hopefully it addresses your earlier "Never to that." review comment! (If it doesn't, could you give me a hint? :wink:) Thanks.

igor-raits commented on main.c:
```diff
@@ -612,6 +612,63 @@ apply_excludes (Pool       *pool,
+  if (ch == NULL)
+  {
+    g_printerr ("Unable to open file '%s': %s\n", filename, err->message);
+    exit (EXIT_FAILURE);
```
what I meant is that function should never do `exit()`. You should set error and return FALSE. So you should make function to return bool

igor-raits commented on main.c:
```diff
@@ -612,6 +612,63 @@ apply_excludes (Pool       *pool,
+                                  int         flags,
+                                  Queue      *solvq)
+{
+  g_autoptr(GError) err = NULL;
```
this error should be part of function declaration so that caller can set it

igor-raits commented on main.c:
```diff
@@ -788,18 +845,11 @@ main (int   argc,
-        queue_push (&pile, q.elements[j]);
+      // solvables prefixed by @ are file names
+      if (**solvable == '@')
+        add_solvables_from_file_to_queue(*solvable+1, pool, &sel, sel_flags, &pile);
```
space before parenthesis

igor-raits commented on main.c:
```diff
@@ -788,18 +845,11 @@ main (int   argc,
+      if (**solvable == '@')
+        add_solvables_from_file_to_queue(*solvable+1, pool, &sel, sel_flags, &pile);
+      else
+        add_solvable_to_queue(*solvable, pool, &sel, sel_flags, &pile);
```
same here

mmathesius commented:
@ignatenkobrain Thank you for your earlier feedback. Apologies for the long delay making the requested changes--I was away.

In addition to the changes you requested, I also: 
* rebased to latest master branch
* renamed subroutines to be consistent with other subroutines in latest master
* rather than having `add_solvables_from_file_to_pile()` return a boolean as earlier suggested, it is cleaner to simply check if the new parameter `error` has been set

r4f4 commented on main.c:
```diff
@@ -831,18 +893,18 @@ pile_from_solvables (Pool  *pool,
         {
-          g_warning ("Nothing matches '%s'", *solvable);
-          continue;
+          queue_init (&pile);
```
Shouldn't this be queue_free instead of queue_init?

mmathesius replied:
Since this function is returning a `Queue`, it's probably best not to free it before returning it. But it really doesn't matter since the caller will be exiting immediately because this is in the midst of returning an error.

Although, perhaps, this function should be adjusted so the caller instead passes in `pile` as a `Queue` pointer?

r4f4 commented on main.c:
```diff
@@ -819,9 +819,71 @@ resolve_all_solvables (Pool  *pool,
+static void
+add_solvable_to_pile (const char *solvable,
+                      Pool       *pool,
+                      Queue      *selection,
```
If you create the selection inside this function, you don't have to pass it to both `add_solvable_to_pile` and `add_solvables_from_file_to_pile` (where it's not even used other than to pass it down to `add_solvable_to_pile`).

mmathesius replied:
Nice catch. I'll take care of that.

igor-raits commented:
I will look at this ASAP.

On Thu, Sep 13, 2018, 23:36 Merlin Mathesius <notifications@github.com>
wrote:

> @ignatenkobrain <https://github.com/ignatenkobrain> Thank you for your
> earlier feedback. Apologies for the long delay making the requested
> changes--I was away.
>
> In addition to the changes you requested, I also:
>
>    - rebased to latest master branch
>    - renamed subroutines to be consistent with other subroutines in
>    latest master
>    - rather than having add_solvables_from_file_to_pile() return a
>    boolean as earlier suggested, it is cleaner to simply check if the new
>    parameter error has been set
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/fedora-modularity/fus/pull/26#issuecomment-421161275>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACu-rkUnuq5RpxljUbIuv6giOYwwQlKsks5uas_ggaJpZM4WLvzE>
> .
>
-- 

-Igor Gnatenko

igor-raits commented on main.c:
```diff
@@ -819,9 +819,71 @@ resolve_all_solvables (Pool  *pool,
+  GIOChannel *ch = g_io_channel_new_file (filename, "r", error);
+  if (ch == NULL)
+  {
+    g_prefix_error (error, "Unable to open file '%s': ", filename);
```
are you sure we need this?

mmathesius replied:
Without it, the message reported is `Error: No such file or directory`--with gives the user no indication of the troublesome file name. With it, the message is reported as `Error: Unable to open file 'nosuchfilefilename': No such file or directory`.

igor-raits commented on main.c:
```diff
@@ -819,9 +819,71 @@ resolve_all_solvables (Pool  *pool,
+
+  GIOStatus ret;
+  do
+  {
```
style

igor-raits commented on main.c:
```diff
@@ -819,9 +819,71 @@ resolve_all_solvables (Pool  *pool,
+{
+  GIOChannel *ch = g_io_channel_new_file (filename, "r", error);
+  if (ch == NULL)
+  {
```
style

mmathesius replied:
For my future reference, is there a guide somewhere to the target style you are looking for? Thanks.

mmathesius commented:
@ignatenkobrain @r4f4 Thanks for your helpful feedback. All of your previously mentioned concerns should be addressed now. I also rebased to the latest master branch.

r4f4 commented on main.c:
```diff
@@ -819,33 +819,89 @@ resolve_all_solvables (Pool  *pool,
+
+      ret = g_io_channel_read_line (ch, &content, &length, &tpos, error);
+      if (ret != G_IO_STATUS_NORMAL || content == NULL)
+	continue;
```
Missing indentation.

r4f4 commented on main.c:
```diff
@@ -819,33 +819,89 @@ resolve_all_solvables (Pool  *pool,
+
+      content[tpos] = '\0';
+      if (*content)
+	add_solvable_to_pile (content, pool, pile);
```
Missing indentation.

mmathesius replied:
Somehow my editor slipped in a TAB instead of 8 spaces in these two spots. Argh!!

r4f4 commented:
Apart from 2 minor indentation issues, LGTM.

igor-raits commented on main.c:
```diff
@@ -819,33 +819,89 @@ resolve_all_solvables (Pool  *pool,
+                     GStrv    solvables,
+                     GError **error)
+{
+  g_clear_error (error);
```
no, you should use return_if_fail if error is not NULL, https://developer.gnome.org/glib/stable/glib-Error-Reporting.html

igor-raits commented on main.c:
```diff
@@ -819,33 +819,89 @@ resolve_all_solvables (Pool  *pool,
-        queue_push (&pile, q.elements[j]);
+      /* solvables prefixed by @ are file names */
+      if (**solvable == '@')
+        add_solvables_from_file_to_pile (*solvable+1, pool, pile, error);
```
spaces around `+`

igor-raits commented on main.c:
```diff
@@ -819,33 +819,89 @@ resolve_all_solvables (Pool  *pool,
+  g_clear_error (error);
+}
+
+static gboolean
```
make this return Queue *, which would be NULL in case of error

mmathesius committed:
Enable ability to read solvables from file by prefixing name with @

Fixes: https://github.com/fedora-modularity/fus/issues/24

Signed-off-by: Merlin Mathesius <mmathesi@redhat.com>

mmathesius commented:
@ignatenkobrain @r4f4 Thanks again for your feedback. Is this getting closer?

And, where can I find the specification of the very exacting coding style that is expected for this? I'd find it very helpful.

igor-raits commented:
https://www.gnu.org/prep/standards/html_node/Formatting.html#Formatting

With a bit different function declaration style and spaces instead of tabs. I think it's all differences

mmathesius commented:
@ignatenkobrain Thank you for the pointer to the style guide. I'll try to be more diligent in sticking to it.

r4f4 commented on main.c:
```diff
@@ -819,33 +820,91 @@ resolve_all_solvables (Pool  *pool,
-  Queue pile;
-  queue_init (&pile);
   int sel_flags = SELECTION_NAME | SELECTION_PROVIDES | SELECTION_GLOB |
                   SELECTION_CANON | SELECTION_DOTARCH;
```
We could make `sel_flags` `static` so that it's initialized only once for the lifetime of the program, not on every call to `add_solvable_to_pile`.

igor-raits replied:
agree

r4f4 commented on main.c:
```diff
@@ -819,33 +820,91 @@ resolve_all_solvables (Pool  *pool,
+                     GStrv    solvables,
+                     GError **error)
+{
+  g_return_val_if_fail (*error == NULL, NULL);
```
Is this really needed? If there was an error before we would've already exited the program.

igor-raits replied:
`g_return_val_if_fail (!error || !*error, FALSE);`

r4f4 commented on main.c:
```diff
@@ -819,33 +820,91 @@ resolve_all_solvables (Pool  *pool,
+  GIOChannel *ch = g_io_channel_new_file (filename, "r", error);
+  if (ch == NULL)
+    {
+      g_prefix_error (error, "Unable to open file '%s': ", filename);
```
Maybe just prefixing the `filename` is enough?

igor-raits commented on main.c:
```diff
@@ -819,33 +820,91 @@ resolve_all_solvables (Pool  *pool,
+      return;
+    }
+
+  /* make sure any error is cleared to indicate success */
```
this sounds very wrong. never clean up errors.

mmathesius replied:
I included that because it's unclear if `error` is guaranteed to be unset upon EOF.

igor-raits commented on main.c:
```diff
@@ -819,33 +820,91 @@ resolve_all_solvables (Pool  *pool,
+    queue_push (pile, q.elements[j]);
+}
+
+static void
```
this should be `static gboolean` which would indicate failure/success

igor-raits commented on main.c:
```diff
@@ -819,33 +820,91 @@ resolve_all_solvables (Pool  *pool,
     }
 
-  return pile;
+  return g_steal_pointer (&pile);
```
does it really work? I always used `g_malloc0()` to create pointer.

If it really does and valgrind is happy - I'm fine.

igor-raits commented on main.c:
```diff
@@ -990,23 +1049,25 @@ main (int   argc,
 
-  g_auto(Queue) pile = pile_from_solvables (pool, solvables);
-  if (!pile.count)
+  g_autoptr(Queue) pile = pile_from_solvables (pool, solvables, &err);
```
hm, do we want to make it `Queue *` here?

mmathesius replied:
I'm not sure what you're asking for. I changed `pile_from_solvables()` to return `Queue *` as requested previously, and `g_autoptr(Queue)` is `Queue *`.  Do you want it changed back to `Queue`?

mmathesius committed:
Updates based on review feedback.

Fixes: https://github.com/fedora-modularity/fus/issues/24

Signed-off-by: Merlin Mathesius <mmathesi@redhat.com>

mmathesius commented:
I pushed an additional commit instead of amending my previous commit to more easily track updates to this PR. Please squash the commits if and when this ever gets merged.

`g_steal_pointer()` is being used exactly as advertised for its intended use. However, I did change the allocation to use `g_malloc0()` instead of `g_malloc()` as suggested. Valgrind shows a total of 32 bytes lost from within `pile_from_solvables()`--but that's a whole lot better than the 95k it shows being definitely lost elsewhere!

Other changes have been made as requested.

Thank you again for all the feedback.

r4f4 commented on main.c:
```diff
@@ -874,34 +874,35 @@ add_solvables_from_file_to_pile (const char  *filename,
 
-  /* make sure any error is cleared to indicate success */
-  g_clear_error (error);
+  g_io_channel_unref (ch);
```
Move this unref to just before the last `if` otherwise you leak `ch` if `ret != G_IO_STATUS_NORMAL`.

r4f4 replied:
Or just make it a `g_autoptr` and remove the unref.

r4f4 commented:
Those 32 bytes being definitely lost are the result of the `g_malloc0`. You need to free that.
The autoptr cleanup function you set for `Queue` is `queue_free` but that only frees the queue's contents, not the pointer you allocated to hold the queue.

r4f4 commented:
@mmathesius do you have a valgrind log of those 95k definitely lost you encountered?

mmathesius commented:
> @mmathesius do you have a valgrind log of those 95k definitely lost you encountered?

https://paste.fedoraproject.org/paste/YM2tDWI1ftoZQAJ2YxLlhg

r4f4 commented:
Why can't we just create the pile before calling the function and pass it as an argument? That way `pile_from_solvables` doesn't have to deal with the pile's scope. We could even rename it to `gboolean add_solvables_to_pile (Pool *pool, Queue *pile, GStr solvables, GError **error)`.
@ignatenkobrain ^

mmathesius commented:
> Why can't we just create the pile before calling the function and pass it as an argument? That way `pile_from_solvables` doesn't have to deal with the pile's scope. We could even rename it to `gboolean add_solvables_to_pile (Pool *pool, Queue *pile, GStr solvables, GError **error)`.
> @ignatenkobrain ^

I had it that way in a previous revision, but was asked to change it...

r4f4 commented:
https://paste.fedoraproject.org/paste/D6v9bnhpUzG5Sk8NC37q6Q

r4f4 commented:
@mmathesius would you mind running your valgrind line again with the following changes? https://paste.fedoraproject.org/paste/JVdLuSimHhCMnjv5Tvpqbw

r4f4 commented:
I created a new issue #42 to track the memleaks.

mmathesius committed:
More updates based on review feedback.

Fixes: https://github.com/fedora-modularity/fus/issues/24

Signed-off-by: Merlin Mathesius <mmathesi@redhat.com>

mmathesius commented:
> https://paste.fedoraproject.org/paste/D6v9bnhpUzG5Sk8NC37q6Q

I added another commit to this PR that corresponds to the above patch. (Thanks--we were thinking alike.) It definitely helps to revert to passing the pile as an argument.

> https://paste.fedoraproject.org/paste/JVdLuSimHhCMnjv5Tvpqbw

Here is the valgrind output that includes the above patch (on top of the commit with the previous patch): https://paste.fedoraproject.org/paste/FjC2tB5s08n3OK1xjlmwqw

That has cut back on the definitely lost memory, but introduces a few new "Invalid free..." reports.

r4f4 commented:
@mmathesius thank you. I think I know the cause and what to do now.

r4f4 commented on main.c:
```diff
@@ -17,6 +17,7 @@
 #include <solv/solv_xfopen.h>
 
 G_DEFINE_AUTOPTR_CLEANUP_FUNC(Pool, pool_free);
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(Queue, queue_free);
```
This is not being used and is in fact wrong, since queue_free does not free the pointer itself.

r4f4 commented:
@ignatenkobrain except for the one line I commented on, I think this PR is ok to go. What do you think?

mmathesius committed:
Another update based on review feedback.

Fixes: https://github.com/fedora-modularity/fus/issues/24

Signed-off-by: Merlin Mathesius <mmathesi@redhat.com>

mmathesius commented:
@ignatenkobrain @r4f4 Do you have any more feedback on this PR, or is it finally ready to be (squashed and) merged?

r4f4 commented:
LGTM

igor-raits reviewed (approved):
Sorry for long delay, was busy week.
